pkg/zcapld/verify.go

```diff
@@ -0,0 +1,193 @@
+	if err != nil {
+		return fmt.Errorf("invalid capability chain: %w", err)
+	}
+
+	// 3. verify the invoker...
+	// authorized invoker must match the verification method itself OR
+	// the controller of the verification method
+	isInvoker, err := isInvoker(proof.Capability, invocation.VerificationMethod)
+	if err != nil {
+		return fmt.Errorf("isInvoke: %w", err)
+	}
+
+	if !isInvoker {
+		return errors.New("the authorized invoker does not match the verification method or its controller")
+	}
+
+	// Begin ControllerProofPurpose
+
+	// TODO the code here validates the proof's "created" time against an expected time:
+	//  nolint:lll // don't want to break the link in two
+	//  https://github.com/digitalbazaar/jsonld-signatures/blob/8d91bcb351702dde4863fab660d7ca1e5e90b2a2/lib/purposes/ProofPurpose.js#L49-L57.
+	//  Note: the higher layers in bedrock-edv-storage don't actually set `proof.created`, so `created` is always NaN.
+	//  Do we really need to verify the proof's date at this layer though? Isn't that the responsibility of a higher
+	//  layer, ie the one that parses and verifies the http signature?
+
+	// TODO verify authorization of verificationMethod.ID by controller for proof purpose `capabilityInvocation`.
+	//  Controller are probably DIDs. They have a "capabilityInvocation" property (just like DIDs) that has
+	//  verificationMethod IDs.
+
+	return nil
+}
+
+// nolint:funlen,gocyclo // TODO decompose verifyCapabilityChain into smaller units
+func (v *Verifier) verifyCapabilityChain(
+	capability *Capability, intendedAction string, invocation *CapabilityInvocation) error {
+	// 1.1. Ensure `capabilityAction`, if given, is allowed; if the capability
+	// restricts the actions via `allowedAction` then it must be in the set.
+	if len(capability.AllowedAction) > 0 && intendedAction != "" &&
+		!stringsContain(capability.AllowedAction, intendedAction) {
+		return fmt.Errorf(
+			`capability action "%s" is not allowed by the capability; allowed actions are: %+v`,
+			intendedAction, capability.AllowedAction)
+	}
+
+	if invocation.ExpectedAction != intendedAction {
+		return fmt.Errorf(
+			`capability action "%s" does not match the expected capability action of "%s"`,
+			intendedAction, invocation.ExpectedAction)
+	}
+
+	// 3. Validate the capability delegation chain.
+	err := capability.validateCapabilityChain()
+	if err != nil {
+		return fmt.Errorf("invalid capability chain: %w", err)
+	}
+
+	// 2. Get the capability delegation chain for the capability.
+	capabilityChain, err := capability.capabilityChain()
+	if err != nil {
+		return fmt.Errorf("failed to fetch capabilityChain: %w", err)
+	}
+
+	// 4. Verify root capability (note: it must *always* be dereferenced since
+	// it does not need to have a delegation proof to vouch for its authenticity
+	// ... dereferencing it prevents adversaries from submitting an invalid
+	// root capability that is accepted):
+	isRoot := len(capabilityChain) == 0
```

Comment:
Your comment numbering is out of order ( 3, 2, 4?)

Reply:
Yes, it results from my implementation following a slightly different order than the reference implementation